Branch

danielbayevski opened the pull request.

danielbayevski committed:
example

danielbayevski committed:
example

danielbayevski committed:
delete all

danielbayevski reviewed (commented):
מה עם זה?

danielbayevski committed:
push stuff

danielbayevski committed:
push stuff

danielbayevski reviewed (commented):
decide_row_num_halt - just to decide how many line we should work with

added  get_time_replace_reduct_cols with a check for discrepencies (OTP and random numbers)

forgot to write changes into the name of the commit

danielbayevski committed:
i realized there ia a problem, i cannot guarantee all cathegories will be in the dictionray
    nor can i guarantee all points (used in KMedioids) will be in the first interation
    therefore two options: use first iteration and pray all groups are represented
    or find all mediod points for each iteration, then find the mediod for each group
    it can be done though with much smaller groups

danielbayevski committed:
two big changes:

A.had to solve the problem of the edge case: "what if
there are no points from each of the groupings in the
first iteration":
solution: find medoids for each iteration then find medoids
 for medoids of all iterations.
thus, we ensure finding the best medoids without doing the
operation over all the input

that leads to inefficiencies:
augmenting the input twice - once for finding model,
the second for prediction. augmentation takes a long
time if columns have a mix of numbers and other objects.

not sure what to do about it.

B.had to rework how replacement dictionary worked, first
 go column by column, save all unique values.
then if there is a mix of numbers and objects - if there
are more than 80% numbers - dictionary all non-object
numbers to zeros, and numbers to floats (because they are stored as str or objects).
this creates a very big dictionary.

AShugaley commented on main.py:
```diff
@@ -1,19 +1,18 @@
-#     row_num= DATABUFFER/ sys.getsizeof(temp)
 
 
 def load_data(path,startrows=None,endrows=None,cols=None):
```
yeah.... No. 

Every function should be no more than 4 lines. And I mean it. EVERY single function. FOUR lines AT MOST (excl the def). NO EXCEPTIONS.

FAQ:
q: Why?
a: The long answer is because this is more readble, modular etc. The short answer: this is they way absolutely everyone does it and so there's really no point in trying to prove something else.

q: Ok I'll do it next time 
a: No, do it this time. Don't create tech debt. 

q: But what about this <insert whatever case>, you see because <insert reason>, this must be an exception!
a: NO EXCEPTIONS!!!!! FOUR LINES!!!!!!!!!

AShugaley replied:
Also, read this: https://github.com/jnguyen095/clean-code/blob/master/Clean.Code.A.Handbook.of.Agile.Software.Craftsmanship.pdf

AShugaley commented on main.py:
```diff
@@ -22,10 +21,19 @@ def load_data(path,startrows=None,endrows=None,cols=None):
-        temp=pd.read_csv(path,usecols=cols,nrows=endrows-startrows,skiprows=startrows)
+    row_num = endrows-startrows if startrows else endrows
+    try:
+        if path[-5:] == ".xlsx" or path[-4:] == ".xls":
```
Surely there must be a nicer way to do this

AShugaley commented on main.py:
```diff
@@ -22,10 +21,19 @@ def load_data(path,startrows=None,endrows=None,cols=None):
+            header =pd.read_excel(path,usecols=cols,nrows=1)
+            temp = pd.read_excel(path,usecols=cols,nrows=row_num,skiprows=startrows)
+            temp.columns = header.columns
+        elif path[-4:] == ".csv":
```
And this

AShugaley commented on main.py:
```diff
@@ -22,10 +21,19 @@ def load_data(path,startrows=None,endrows=None,cols=None):
+            temp.columns = header.columns
+        elif path[-4:] == ".csv":
+            header = pd.read_csv(path, usecols=cols, nrows=1)
+            temp=pd.read_csv(path,usecols=cols,nrows=row_num,skiprows=startrows)
```
[STYLE] space after the `=`

AShugaley replied:
Please read whatever style guide for Python you can find, for example this: https://peps.python.org/pep-0008/ and follow it.

Also, do you work with an IDE? If yes - make it check those kinds of stuff. If no - start working with an IDE

AShugaley commented on main.py:
```diff
@@ -35,91 +43,200 @@ def save_file(path, pred, startrows=None, endrows=None):
+    timeCols = []
+    replaceDict = {}
+    colNums = []
+    numberTypes = numpy.array(["int8","int16","int32","int64","uint8","uint16","uint64","uint32","float32","float64"])
```
Surely there must be a nicer way. 

But if there isn't (there is) - at least make it a constant

AShugaley commented on main.py:
```diff
@@ -35,91 +43,200 @@ def save_file(path, pred, startrows=None, endrows=None):
+                        replaceDict.update({col:  dictionary.copy()})
+                data1 = load_data(path, startrows=rowsForSingleColumn * p, endrows=rowsForSingleColumn * (1 + p),
+                                 cols=[i])
+                p += 1
```
Was this 4 lines? I don't think so.

AShugaley commented on main.py:
```diff
@@ -35,91 +43,200 @@ def save_file(path, pred, startrows=None, endrows=None):
+        save_file(path, prediction, row_num * (i), row_num * (i + 1))
+        data = load_data(path, row_num * (i + 1), row_num * (i + 2))
+        i += 1
+    print("save: " + str(time.time() - t))
```
You know what I want to say.

AShugaley reviewed (changes requested):
I deliberately made zero comments about the logic itself. The truth is - the logic doesn't matter.

This is something that junior engineers are really surprised to know. They come, they write code, the code works - but the other engineers simply completely reject it and demand it to be rewritten completely. "But my code WORKS!!!" it WORKS!!! Rewriting is such a waste of time!!!" cry the junior engineers. 

And yet, no one cares. Any code that you'll eventually push will work (otherwise you won't push it), PRs don't really check this (you check this with tests. Which reminds me - write tests). PRs check if this is a clean code. 

So that's the plan:
1. Clean this up - follow a style guide and Clean Code principles 
2. Add tests
3. Then we'll discuss the logic itself

P.S. Seriously - write tests. Find some recommended testing library for Py, learn how to use it and add tests. If you want to learn - do it the right way. 
P.S.S Sorry, I know this is super annoying. It was annoying for me as well. It was annoying for any engineer ever that ever lived. That's simply the way things work

danielbayevski committed:
had problems with preprocessing, later
1.tidies up the code, 4 lines per function; any '=' sign has sapce, and so forth
2. added a test for the whole deal, will add unit tests later.
will find more CSV,XLS file later as well, for testing

currently, 2 tests have 85% success rate. 4 range between 20-50% those try to categorize more obscure things, it seems that k-medoids is not a good algorithm for them, if its even possible

also, preprocessing is kinda weird, lets discuss when you're available

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+                  endRows = None, cols = None, 
+                  sheet = 0):
+    header = pd.read_excel(path, usecols = cols, nrows = 1, sheet_name = sheet)
+    temp = pd.read_excel(path,
```
Huh? Why do you access the file twice? Not sure I understand what's going on here

danielbayevski replied:
if i dont take the headers (the first line, the titles) then the second read_excel doesnt take the column headers into account and i get columns with title "75.12". i later use the column names

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+                       skiprows = startRows,
+                       thousands = ",")
+    temp.columns = header.columns
+    return temp
```
This function and the other one are exactly the same. Don't duplicate code.

```
def load_data_excel(path, ....):
    load data(path, ...., read, pd.read_excel)
    
def load_data_csv(path, ....):
    load data(path, ...., read, pd.read_csv)
    
def load_datal(path, ...., read_func):
    header = read_func(path, usecols = cols, nrows = 1)
    ...
```

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+                   endRows = None,
+                   cols = None,
+                   sheet = 0):
+    if path[-5:] == ".xlsx" or path[-4:] == ".xls":
```
please split this into a ton of functions
```
import pathlib

EXCEL_FILES = ['.xls', '.xlsx']
EXCEL_FILES = ['.csv']

def get_file_type(path):
    pathlib.Path(path).suffix

def is_excel(path)
    get_file_type(path) in EXCEL_FILES

def is_csv(path)
    get_file_type(path) in CSV_FILES
   
def load_data_type(...)
    if excel_file(path) return load_data_xls(path, startRows, endRows, cols, sheet)
    if csv_file(path) load_data_csv(path, startRows, endRows, cols)
    # throw an "unsupported file" exception here! and check for this with a cute test - for all three outcomes
    ```

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+    try:
+        return load_data_type(path, startRows, endRows, cols, sheet)
+    except:
+        return pd.DataFrame()
```
Not sure what this function does

danielbayevski replied:
this function is to catch the load when it gets a startRow higher than the amount of rows it has, it then returns an empty dataFrame.
i could do it nicer
what im trying to avoid is loading too many lines (for example with a csv with a 10 million lines and overloading the memory)

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+
+
+def save_file_csv(path, temp, newDirectory):
+    new_path = newDirectory+path[-(len(path)-path.rfind("/")):]
```
code duplication here, move to a shared function (as I showed above)

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+
+def save_file_xls(path, temp, sheet, newDirectory):
+    new_path = newDirectory+path[-(len(path)-path.rfind("/")):]
+    save_file_xls2(new_path, temp, sheet)
```
What? haha

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
     if path[-5:] == ".xlsx" or path[-4:] == ".xls":
-        temp = pd.read_excel(path,usecols=cols,nrows=endrows-startrows,skiprows=startrows)
+        save_file_xls(path, temp, sheet, newDirectory)
     elif path[-4:] == ".csv":
```
Once you add the "get file type" function - use it here and everywhere

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
-        new_path = path[:-4] + "_new" + path[-4:]
-        temp = pd.concat(temp, pred)
-        temp.to_csv(new_path)
+    temp = load_data(path, startRows, endRows, sheet = sheet)
```
Don't use `temp` as a variable name. Ever.

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
-        temp = pd.concat(temp, pred)
-        temp.to_csv(new_path)
+    temp = load_data(path, startRows, endRows, sheet = sheet)
+    temp["grouping"] = prediction.tolist()
```
Not sure what's going on here. Why do you need this? This code is very hard to read. I understand that some of it has to do with the specific library you use to handle the excel / csv. But you need to think about a way to make it easier and clearer. 

Also - use constants for everything that's constant (e.g. `"grouping"`)

danielbayevski replied:
im re-loading the data, joining it with the predictions then saving it.
i could just save a vector of all the predictions, but i think it would be more useful with all the data and calling that column "grouping"
i have to re-load the data because the data i use gets preprocessed
i could preprocess a copy of the data and lose the re-loading, but it would mean i have to use more memory (exactly what im trying to avoid)

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+                    {column_name: LabelEncoder().fit(list(val_set))})
 
+    single_col_row_num = decide_row_num_halt(path, [real_number_cols[0]])
+    mean_std = {}
```
Too long

danielbayevski replied:
you mean the function name?

AShugaley commented on main.py:
```diff
@@ -1,132 +1,379 @@
+                  sheet = new_sheet)
+        i += 1
+        data = load_data(path, row_num * i, row_num * (i + 1),
+                         cols = colNum, sheet = sheet)
```
Way too long.

AShugaley reviewed (changes requested):
OK, this is better. 

A few changes:
1. It's time to split this into multiple files. Two options:
      a. Use OOP. So create a class for every potential object (e.g. `excelReader` or whatever). This is better but will require a lot of changes
       b. Use modules. This is easier to do here since it's mostly just copy pasting functions from this file to other files. https://www.analyticsvidhya.com/blog/2021/07/working-with-modules-in-python-must-known-fundamentals-for-data-scientists/#:~:text=In%20Python%2C%20Modules%20are%20simply,to%20include%20in%20your%20application. The idea is to group them by context - all function that handle reading and writing should be in one module. And this module should include three nested modules - excel, csv and shared or general. Again, this is better and easier to do with classes since it seems that Python does not let you extend modules (correct me if I'm wrong). But both options are fine with me.


2. Add tests. And by tests I'll start from unit tests that check specific small functions. For example you have a function that received a path and if it's an excel file. Write two small tests (pass and fail) to check this function only. Then go to a function that receives a path and calls either read_excel or read_csv - check that it calls the right function (don't check the functionality of the read_csv - just check that it's been called).

3. 4 lines. 4 lines.
